🔥 add firestore recipe 🔥 (#3223)
withastro/docs: src/content/docs/en/guides/backend/google-firebase.mdx

The endpoint will verify the ID token and create a new session cookie for the user. Then, the endpoint will redirect the user to the `/dashboard` page.

## Connecting to Firestore database

### Prerequisites

- An Astro project initialized with Firebase as described in the [Initializing Firebase in Astro](#initializing-firebase-in-astro) section.

- A Firebase project with a Firestore database. You can follow the [Firebase documentation](https://firebase.google.com/docs/firestore/quickstart) to create a new project and set up a Firestore database. 

> Our Firestore collection will be called **friends** and will contain documents with the following fields:

Reviewer: ```suggestion
In this recipe, the Firestore collection will be called **friends** and will contain documents with the following fields:
```

- `id`: autogenerated by Firestore
- `name`: a string field
- `age`: a number field
- `isBestFriend`: a boolean field

> Feel free to follow this example or adapt it to your own Firestore collection.

Reviewer: ```suggestion
```

### Creating the server endpoints

> We will create two endpoints to interact with the Firestore database:
> 
> - `POST /api/friends`: to create a new document in the friends collection.
> - `GET /api/friends/:id`: to get a document in the friends collection.
> - `POST /api/friends/:id`: to update a document in the friends collection.
> - `DELETE /api/friends/:id`: to delete a document in the friends collection.
> 
> Create two new files in a new directory `src/pages/api/friends/`: `index.ts` and `[id].ts`.

Reviewer: ```suggestion
Create two new files in a new directory `src/pages/api/friends/`: `index.ts` and `[id].ts`. These will create two server endpoints to interact with the Firestore database in the following ways:

- `POST /api/friends`: to create a new document in the friends collection.
- `GET /api/friends/:id`: to get a document in the friends collection.
- `POST /api/friends/:id`: to update a document in the friends collection.
- `DELETE /api/friends/:id`: to delete a document in the friends collection.

```

    return new Response("Missing required fields", {
      status: 400,
    });
  }

  if (!params.id) {
    return new Response("Cannot find friend", {
      status: 404,
    });
  }

  try {
    await friendsRef.doc(params.id).update({
      name,
      age: parseInt(age),
      isBestFriend,
    });
  } catch (error) {
    console.error(error);
  }
  return redirect("/dashboard");
};

export const del: APIRoute = async ({ params, redirect }) => {
  if (!params.id) {
    return new Response("Cannot find friend", {
      status: 404,
    });
  }

  try {
    await friendsRef.doc(params.id).delete();
  } catch (error) {
    console.error(error);
  }
  return redirect("/dashboard");
}
```

:::note
This is a basic implementation of the `friends/:id` endpoint. You can add more logic to this endpoint to suit your needs.
:::

After creating server endpoints for Firestore, your project directory should now include these new files:

<FileTree title="Project Structure">
- src
  - env.d.ts
  - firebase
    - client.ts
    - server.ts
  - pages
    - api
      - friends
        - **index.ts**
        - **[id].ts**
- .env
- astro.config.mjs
- package.json
</FileTree>

### Creating pages

> Now that we have Firestore endpoints set up, we can create the pages that will use them. 

Reviewer: ```suggestion
Create the pages that will use the Firestore endpoints:
```

- `src/pages/add.astro` - will contain a form to add a new friend.
- `src/pages/edit/[id].ts` - will contain a form to edit a friend and a button to delete a friend.
- `src/pages/friend/[id].ts` - will contain the details of a friend.
- `src/pages/dashboard.astro` - will display a list of friends.

> The example `src/pages/add.astro` below includes a form that will send a `POST` request to the `/api/friends` endpoint. This endpoint will create a new friend using the data from the form and then will redirect the user to the `/dashboard` page.

Reviewer: ```suggestion
#### Add a new record

The example `src/pages/add.astro` below includes a form that will send a `POST` request to the `/api/friends` endpoint. This endpoint will create a new friend using the data from the form and then will redirect the user to the `/dashboard` page.
```

```astro title="src/pages/add.astro"
---
import Layout from "../layouts/Layout.astro";
---

<Layout title="Add a new friend">
  <h1>Add a new friend</h1>
  <form method="post" action="/api/friends">
    <label for="name">Name</label>
    <input type="text" id="name" name="name" />
    <label for="age">Age</label>
    <input type="number" id="age" name="age" />
    <label for="isBestFriend">Is best friend?</label>
    <input type="checkbox" id="isBestFriend" name="isBestFriend" />
    <button type="submit">Add friend</button>
  </form>
</Layout>
```

> `src/pages/edit/[id].astro` will contain a form to edit a friend data and a button to delete a friend. On submit, this page will send a `POST` request to the `/api/friends/:id` endpoint to update a friend data. 

Reviewer: ```suggestion
#### Edit or Delete a record

`src/pages/edit/[id].astro` will contain a form to edit a friend data and a button to delete a friend. On submit, this page will send a `POST` request to the `/api/friends/:id` endpoint to update a friend data. 
```

  const url = document.querySelector('form')?.getAttribute('action') as string
  deleteButton.addEventListener('click', async () => {
    const response = await fetch(url, {
      method: 'DELETE'
    })
    if (response.redirected) {
      window.location.assign(response.url)
    }
  })
</script>
```

`src/pages/friend/[id].astro` will display the details of a friend.

```astro title="src/pages/friend/[id].astro"
---
import Layout from '../../layouts/Layout.astro'
import { app } from '../../firebase/server'
import { getFirestore } from 'firebase-admin/firestore'

interface Friend {
  name: string;
  age: number;
  isBestFriend: boolean;
}

const { id } = Astro.params

if (!id) {
  return Astro.redirect('/404')
}

const db = getFirestore(app)
const friendsRef = db.collection("friends");
const friendSnapshot = await friendsRef.doc(id).get()

if (!friendSnapshot.exists) {
  return Astro.redirect('/404')
}

const friend = friendSnapshot.data() as Friend
---
<Layout title={friend.name}>
  <h1>{friend.name}</h1>
  <p>Age: {friend.age}</p>
  <p>Is best friend: {friend.isBestFriend ? 'Yes' : 'No'}</p>
</Layout>
```

> Finally, `src/pages/dashboard.astro` will display a list of friends. Each friend will have a link to their details page and an edit button that will redirect the user to the edit page.

Reviewer: ```suggestion
#### Display records with an edit button

Finally, `src/pages/dashboard.astro` will display a list of friends. Each friend will have a link to their details page and an edit button that will redirect the user to the edit page.
```

Create the pages that will use the Firestore endpoints:

- `src/pages/add.astro` - will contain a form to add a new friend.
- `src/pages/edit/[id].ts` - will contain a form to edit a friend and a button to delete a friend.
- `src/pages/friend/[id].ts` - will contain the details of a friend.
- `src/pages/dashboard.astro` - will display a list of friends.

#### Add a new record

The example `src/pages/add.astro` below includes a form that will send a `POST` request to the `/api/friends` endpoint. This endpoint will create a new friend using the data from the form and then will redirect the user to the `/dashboard` page.

```astro title="src/pages/add.astro"
---
import Layout from "../layouts/Layout.astro";
---

<Layout title="Add a new friend">
  <h1>Add a new friend</h1>
  <form method="post" action="/api/friends">
    <label for="name">Name</label>
    <input type="text" id="name" name="name" />
    <label for="age">Age</label>
    <input type="number" id="age" name="age" />
    <label for="isBestFriend">Is best friend?</label>
    <input type="checkbox" id="isBestFriend" name="isBestFriend" />
    <button type="submit">Add friend</button>
  </form>
</Layout>
```

> #### Edit or Delete a record

Reviewer: ```suggestion
#### Edit a record
```

if (!friendSnapshot.exists) {
  return Astro.redirect('/404')
}

const friend = friendSnapshot.data() as Friend
---
<Layout title="Edit {friend.name}">
  <h1>Edit {friend.name}</h1>
  <p>Here you can edit or delete your friend's data.</p>
  <form method="post" action={`/api/friends/${id}`}>
    <label for="name">Name</label>
    <input type="text" id="name" name="name" value={friend.name} />
    <label for="age">Age</label>
    <input type="number" id="age" name="age" value={friend.age} />
    <label for="isBestFriend">Is best friend?</label>
    <input type="checkbox" id="isBestFriend" name="isBestFriend" checked={friend.isBestFriend}/>
    <button type="submit">Edit friend</button>
  </form>
  <button type="button" id="delete-document">Delete</button>
</Layout>
<script>
  const deleteButton = document.getElementById('delete-document') as HTMLButtonElement
  const url = document.querySelector('form')?.getAttribute('action') as string
  deleteButton.addEventListener('click', async () => {
    const response = await fetch(url, {
      method: 'DELETE'
    })
    if (response.redirected) {
      window.location.assign(response.url)
    }
  })
</script>
```

> `src/pages/friend/[id].astro` will display the details of a friend.

Reviewer: ```suggestion
#### Display an individual record

`src/pages/friend/[id].astro` will display the details of a friend.
```

  const url = document.querySelector('form')?.getAttribute('action') as string
  deleteButton.addEventListener('click', async () => {
    const response = await fetch(url, {
      method: 'DELETE'
    })
    if (response.redirected) {
      window.location.assign(response.url)
    }
  })
</script>
```

#### Display an individual record

`src/pages/friend/[id].astro` will display the details of a friend.

```astro title="src/pages/friend/[id].astro"
---
import Layout from '../../layouts/Layout.astro'
import { app } from '../../firebase/server'
import { getFirestore } from 'firebase-admin/firestore'

interface Friend {
  name: string;
  age: number;
  isBestFriend: boolean;
}

const { id } = Astro.params

if (!id) {
  return Astro.redirect('/404')
}

const db = getFirestore(app)
const friendsRef = db.collection("friends");
const friendSnapshot = await friendsRef.doc(id).get()

if (!friendSnapshot.exists) {
  return Astro.redirect('/404')
}

const friend = friendSnapshot.data() as Friend
---
<Layout title={friend.name}>
  <h1>{friend.name}</h1>
  <p>Age: {friend.age}</p>
  <p>Is best friend: {friend.isBestFriend ? 'Yes' : 'No'}</p>
</Layout>
```

> #### Display records with an edit button

Reviewer: ```suggestion
#### Display a list of records with an edit button
```
